compiler/rustc_const_eval/src/transform/validate.rs

```diff
@@ -619,6 +619,25 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                             );
                         }
                     }
+                    CastKind::Transmute => {
+                        let src_layout = self.tcx.layout_of(self.param_env.and(op_ty));
+                        let dst_layout = self.tcx.layout_of(self.param_env.and(*target_type));
+                        if let Err(e) = src_layout {
+                            self.fail(
+                                location,
+                                format!("Unable to compute layout for source type {op_ty:?}: {e}"),
+                            );
+                        }
+                        if let Err(e) = dst_layout {
+                            self.fail(location, format!("Unable to compute layout for destination type {target_type:?}: {e}"));
+                        }
+
+                        if let (Ok(src_layout), Ok(dst_layout)) = (src_layout, dst_layout) {
+                            if src_layout.layout.size() != dst_layout.layout.size() {
+                                self.fail(location, format!("Source and destination layouts have different sizes: {src_layout:?} vs {dst_layout:?}"));
+                            }
+                        }
+                    }
```

Comment:
Hmm, thinking about it more, I actually don't think we should have any check like this. Transmutes are presumably things that optimizations want to introduce, and making these kinds of cases not well-formed just feels like a footgun for those opts.

It's probably fine to keep this in for now, but can you include a comment that indicates that we can and should relax this restriction if some opts run into the need for it?

Reply:
Yeah, relaxing it in future sounds reasonable.  I've added the comment, since relaxing it will probably need a bunch of `intrinsics::mir!` tests around how the various consumers would handle such a thing, which I'd rather not do right now.